src/hutch_bunny/core/query_solvers.py

```diff
@@ -207,50 +148,40 @@ def _solve_rules(self) -> None:
                         )
                         .distinct()
                     )
-                    rule_df = pd.read_sql_query(
-                        sql=stmnt, con=self.db_manager.engine.connect()
-                    )
-                    main_df = main_df.merge(
-                        right=rule_df,
-                        how=merge_method(group.rules_operator),
-                        left_on="person_id",
-                        right_on=f"person_id_{i}",
-                    )
-                # Text rules testing for inclusion
-                elif rule.operator == "=":
-                    stmnt = (
-                        select(concept_table.person_id.label(f"person_id_{i}"))
-                        .where(boolean_rule_col == int(rule.value))
-                        .distinct()
-                    )
-                    rule_df = pd.read_sql_query(
-                        sql=stmnt, con=self.db_manager.engine.connect()
-                    )
+                else:
+                    stmnt=self.build_statement(concept_table, boolean_rule_col, rule, label_to_use)
+
+
+                rule_df = pd.read_sql_query(
+                    sql=stmnt, con=self.db_manager.engine.connect()
+                )
+
+                if rule_index>0: #then we merge
                     main_df = main_df.merge(
                         right=rule_df,
                         how=merge_method(group.rules_operator),
                         left_on="person_id",
-                        right_on=f"person_id_{i}",
+                        right_on=f"person_id_{rule_index}",
                     )
-                # Text rules testing for exclusion
-                elif rule.operator == "!=":
-                    stmnt = (
-                        select(concept_table.person_id.label(f"person_id_{i}"))
-                        .where(boolean_rule_col != int(rule.value))
-                        .distinct()
-                    )
-                    rule_df = pd.read_sql_query(
+                else: #then we load the first one
+                    main_df = pd.read_sql_query(
                         sql=stmnt, con=self.db_manager.engine.connect()
                     )
-                    main_df = main_df.merge(
-                        right=rule_df,
-                        how=merge_method(group.rules_operator),
-                        left_on="person_id",
-                        right_on=f"person_id_{i}",
-                    )
+
             # subqueries therefore contain the results for each group within the cohort definition.
             self.subqueries.append(main_df)
 
+    def build_statement(self, concept_table, boolean_rule_col, rule, label):
```

Comment:
Would be helpful to have this typed where we can

https://peps.python.org/pep-0484/
For more information